core/src/main/java/org/apache/brooklyn/util/core/osgi/BundleMaker.java

```diff
@@ -0,0 +1,340 @@
+                @Override
+                public boolean apply(String input) {
+                    return filter.apply(input) && !entryNames.contains(input);
+                }
+            });
+            
+            if (!addAtStart) {
+                writeZipEntries(zout, entries);
+            }
+
+            return f2;
+        } catch (IOException e) {
+            throw Exceptions.propagate("Unable to read "+f+" when looking for manifest", e);
+        } finally {
+            Streams.closeQuietly(zf);
+            Streams.closeQuietly(zout);
+        }
+    }
+
+    private void writeZipEntries(ZipOutputStream zout, Map<ZipEntry, InputStream> entries) throws IOException {
+        for (Map.Entry<ZipEntry,InputStream> ze: entries.entrySet()) {
+            zout.putNextEntry(ze.getKey());
+            InputStream zin = ze.getValue();
+            Streams.copy(zin, zout);
+            Streams.closeQuietly(zin);
+            zout.closeEntry();
+        }
+    }
+
+    private void writeZipEntriesFromFile(ZipOutputStream zout, File existingZip, Predicate<String> filter) throws IOException {
+        ZipFile zf = new ZipFile(existingZip);
+        try {
+            Enumeration<? extends ZipEntry> zfe = zf.entries();
+            while (zfe.hasMoreElements()) {
+                ZipEntry ze = zfe.nextElement();
+                if (filter.apply(ze.getName())) {
+                    zout.putNextEntry(ze);
+                    InputStream zin = zf.getInputStream(ze);
+                    Streams.copy(zin, zout);
+                    Streams.closeQuietly(zin);
+                    zout.closeEntry();
+                }
+            }
+        } finally {
+            Streams.closeQuietly(zf);
+        }
+    }
+    
+    /** installs the given JAR file as an OSGi bundle; all manifest info should be already set up.
+     * bundle-start semantics are TBD. */
+    @Beta
+    public Bundle installBundle(File f, boolean start) {
+        try {
+            Bundle b = Osgis.install( framework, "file://"+f.getAbsolutePath() );
+            if (start) {
+                // benefits of start:
+                // a) we get wiring issues thrown here, and
+                // b) catalog.bom in root will be scanned synchronously here
+                // however drawbacks:
+                // c) other code doesn't always do it (method above)
+                // d) heavier-weight earlier
+                // e) tests in IDE break (but mvn fine)
+                b.start();
+            }
+            
+            return b;
+            
+        } catch (Exception e) {
+            throw Exceptions.propagate("Error starting bundle from "+f, e);
+        }
+    }
+    
+    private boolean addUrlItemRecursively(ZipOutputStream zout, String root, String item, Predicate<String> filter) throws IOException {
+        InputStream itemFound = null;
+        try {
+            itemFound = resources.getResourceFromUrl(item);
```

Comment:
Wow haven't realised it's possible to read the listing of a classpath path. Had to run the test myself to make sure it works :)
Is this standard, documented somewhere? Does it work in Karaf?